source/grid/grid_generator.cc

```diff
@@ -7090,6 +7090,273 @@ namespace GridGenerator
 
 
 
+  template <int dim, int spacedim>
+  void
+  convert_hypercube_to_simplex_mesh(const Triangulation<dim, spacedim> &in_tria,
+                                    Triangulation<dim, spacedim> &out_tria)
+  {
+    AssertDimension(in_tria.n_global_levels(), 1);
+
+    Assert(dim > 1, ExcNotImplemented());
+
+    std::vector<Point<spacedim>> vertices;
+    std::vector<CellData<dim>>   cells;
+    SubCellData                  subcell_data;
+
+    // store for each vertex and face the assigned index so that we only
+    // assign them a value only once
+    std::vector<unsigned int> indices_vertics(in_tria.n_vertices(),
+                                              numbers::invalid_unsigned_int);
+    std::vector<unsigned int> indices_faces(in_tria.n_faces(),
+                                            numbers::invalid_unsigned_int);
+
+    // temporary array storing the global indices of each cell entity in the
+    // sequence: vertices, edges/faces, cell
+    std::array<unsigned int, dim == 2 ? 9 : 14> indices_local;
+
+    // loop over all cells
+    for (const auto &cell : in_tria)
```

Comment:
Here and below, you have a number of comments that just duplicate what the next line does. I can see that you're looping over all cells from the next line, but what I don't know is *why* you're doing that and *what* you are doing this for. A good comment in front of the `for` loop would be something like this:

We first have to create all of the new vertices. To do this, we loop over all cells and on each cell (i) copy the existing vertex locations (and record their new indices in the 'indices_vertics' [replace by new name] array), (ii) create new midpoint vertex locations for each face (and record their new indices in 'indices_faces' [again, replace by new name], (iii) create new midpoint vertex locations for each cell.